pkg/storage/ingest/pusher.go

```diff
@@ -241,163 +253,260 @@ func (c pusherConsumer) shouldLogClientError(ctx context.Context, err error) (bo
+	context.Context
 }
 
-// TODO dimitarvdimitrov if this is expensive, consider having this long-lived and not Close()-ing and recreating it on every fetch, but instead calling something like Flush() on it.
-func newShardingPusher(numTimeSeriesPerFlush prometheus.Histogram, numShards int, batchSize int, buffer int, upstream Pusher) *shardingPusher {
-	pusher := &shardingPusher{
+// newParallelStorageShards creates a new parallelStorageShards instance.
+func newParallelStorageShards(numTimeSeriesPerFlush prometheus.Histogram, numShards int, batchSize int, capacity int, pusher Pusher) *parallelStorageShards {
+	p := &parallelStorageShards{
 		numShards:             numShards,
-		upstream:              upstream,
+		pusher:                pusher,
+		capacity:              capacity,
 		numTimeSeriesPerFlush: numTimeSeriesPerFlush,
 		batchSize:             batchSize,
 		wg:                    &sync.WaitGroup{},
 		errs:                  make(chan error, numShards),
-		unfilledShards:        make([]flushableWriteRequest, numShards),
-	}
-	shards := make([]chan flushableWriteRequest, numShards)
-	pusher.wg.Add(numShards)
-	for i := range shards {
-		pusher.unfilledShards[i].WriteRequest = &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()}
-		shards[i] = make(chan flushableWriteRequest, buffer)
-		go pusher.runShard(shards[i])
 	}
-	go func() {
-		pusher.wg.Wait()
-		close(pusher.errs)
-	}()
 
-	pusher.shards = shards
-	return pusher
+	p.start()
+
+	return p
 }
 
-func (p *shardingPusher) PushToStorage(ctx context.Context, request *mimirpb.WriteRequest) error {
+// ShardWriteRequest hashes each time series in the write requests and sends them to the appropriate shard which is then handled by the current BatchingQueue in that shard.
+func (p *parallelStorageShards) ShardWriteRequest(ctx context.Context, request *mimirpb.WriteRequest) error {
 	var (
 		builder         labels.ScratchBuilder
 		nonCopiedLabels labels.Labels
 		errs            multierror.MultiError
 	)
+
 	for _, ts := range request.Timeseries {
 		mimirpb.FromLabelAdaptersOverwriteLabels(&builder, ts.Labels, &nonCopiedLabels)
 		shard := nonCopiedLabels.Hash() % uint64(p.numShards)
 
-		s := p.unfilledShards[shard]
-		// TODO dimitarvdimitrov support metadata and the rest of the fields; perhaps cut a new request for different values of SkipLabelNameValidation?
-		s.Timeseries = append(s.Timeseries, ts)
-		s.Context = ctx // retain the last context in case we have to flush it when closing shardingPusher
+		// TODO: Add metrics to measure how long are items sitting in the queue before they are flushed.
+		p.shards[shard].AddToBatch(ctx, ts)
 
-		if len(s.Timeseries) < p.batchSize {
-			continue
-		}
-		p.unfilledShards[shard] = flushableWriteRequest{
-			WriteRequest: &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()},
+		// After each addition, check for any errors in the error channel
+		select {
+		case err := <-p.errs:
```

Comment:
This way turned out to be super expensive. I had it like this initially, but replay speed falttened-out because we end up checking the channel about 1M times per second. So I ended up checking this channel once per flush.

Maybe `AddToBatch` can return an error and retain something like the previous behaviour

Reply:
Oh, that's clever - I like it!

Reply:
Please take a look at https://github.com/grafana/mimir/pull/9133/commits/ff3c48c0b5e9ff4991b0a85272858b14ee87cd98 if the approach is OK, I'm going to do a quick scan / cleanup for comments and then follow through with this PR.